lib/attrsets.nix

```diff
@@ -56,9 +56,7 @@ rec {
   */
   setAttrByPath = attrPath: value:
     if attrPath == [] then value
-    else listToAttrs
-      [ { name = head attrPath; value = setAttrByPath (tail attrPath) value; } ];
-
+    else foldr (name: value: { ${name} = value; }) value attrPath;
```

Comment:
GitHub ate my review comment for this line :(, so here's my second attempt.

`foldr` is not a builtin, so this is now causing a lot of allocation of envs (data structure representing the lexical scope). You could try inlining the `foldr` definition and then optimize that. I suspect that may push more if not most of the evaluation stats into the negative percentages.

Reply:
hmm, I didn't consider allocations initially, but it definite makes more sense now. I have updated the implementation to inline `foldr`.

Reply:
I didn't consider the overhead of function calls and envs(I definite have more to learn). I have updated the implementation to inline `foldr`.

Reply:
I wonder if implementing `foldr` with `foldl'` can offer more performance benefits here, as `foldl'` is implemented with iteration. The laziness of `foldr` does not matter given `setAttrByPath` should be atomic and can not short circuit.

Reply:
Some of the list items could be bottom (`abort`, loop, etc) and lazy construction of these sets could avoid it.
It could be a separate function, so call sites can opt into strictness. It seems safe to assume that performance will improve with that.
For performance testing you could replace it of course, but I don't want to risk it, because some user functions may depend on its laziness.

Reply:
I see, the current lazy implementation it is definitely safer in terms of compatibility.